probe-rs/src/probe/espusbjtag/mod.rs

```diff
@@ -35,148 +32,162 @@ pub(crate) struct EspUsbJtag {
+        // TODO this isn't actually needed, we should only do this when AttachUnderReset it supplied
+        self.protocol.set_reset(true, true)?;
+        self.protocol.set_reset(false, false)?;
 
-        tracing::trace!("Response: {:?}", response);
+        self.jtag_reset()?;
 
-        let _remainder = response.split_off(tms_enter_shift.len());
+        let input = Vec::from_iter(iter::repeat(0xFFu8).take(4 * max_chain));
+        let response = self.write_dr(&input, 4 * max_chain * 8).unwrap();
 
-        let mut remaining_bits = register_bits;
+        tracing::trace!("DR: {:?}", response);
 
-        let mut result = Vec::new();
+        let mut idcodes = Vec::new();
 
-        while remaining_bits >= 8 {
-            let byte = bits_to_byte(response.split_off(8)) as u8;
-            result.push(byte);
-            remaining_bits -= 8;
+        for idcode in response.chunks(4) {
+            if idcode.len() != 4 {
+                panic!("Bad length");
+            }
+            if idcode == [0xFF, 0xFF, 0xFF, 0xFF] {
+                break;
+            }
+            idcodes.push(u32::from_le_bytes((idcode).try_into().unwrap()));
         }
 
-        // Handle leftover bytes
-        if remaining_bits > 0 {
-            result.push(bits_to_byte(response.split_off(remaining_bits)) as u8);
-        }
+        tracing::info!(
+            "JTAG dr scan complete, found {} TAPS. {:?}",
+            idcodes.len(),
+            idcodes
+        );
+
+        let input = Vec::from_iter(iter::repeat(0xffu8).take(idcodes.len()));
+        let mut response = self.write_ir(&input, idcodes.len() * 8).unwrap();
+
+        let expected = if let Some(ref chain) = self.scan_chain {
+            let expected = chain
+                .iter()
+                .filter_map(|s| s.ir_len)
+                .map(|s| s as usize)
+                .collect::<Vec<usize>>();
+            response.truncate(expected.iter().sum());
+            Some(expected)
+        } else {
+            None
+        };
 
-        tracing::debug!("Read from DR: {:?}", result);
+        tracing::trace!("ir scan: {}", response.as_bitslice());
 
-        Ok(result)
+        let ir_lens =
+            extract_ir_lengths(response.as_bitslice(), idcodes.len(), expected.as_deref()).unwrap();
+        tracing::trace!("Detected IR lens: {:?}", ir_lens);
+
+        Ok((idcodes, ir_lens))
     }
 
     /// Write IR register with the specified data. The
     /// IR register might have an odd length, so the data
     /// will be truncated to `len` bits. If data has less
     /// than `len` bits, an error will be returned.
-    fn write_ir(&mut self, data: &[u8], len: usize) -> Result<(), DebugProbeError> {
-        if len >= 8 {
-            return Err(DebugProbeError::NotImplemented(
-                "Not yet implemented for IR registers larger than 8 bit",
-            ));
-        }
+    fn write_ir(&mut self, data: &[u8], len: usize) -> Result<BitVec<u8, Lsb0>, DebugProbeError> {
```

Comment:
Don't we have common traits for those nowadays? We can also do that in a followup PR :)

Reply:
I don't think so? If you know where to find them let me know :)

Reply:
I thought some JTAG stuff was unified recently, but I guess not then. Would be a good thing in the future I guess.
But we should hold off on it until we have good HW to test on.